### PR TITLE
Fix mapConcat context propagation

### PR DESCRIPTION
**Pull request by ygree**
Fix mapConcat context propagation, so consecutive child elements are connected to the original span when using Telemetry tracing

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -2230,13 +2236,13 @@ private[akka] final class StatefulMapConcat[In, Out](val f: () => In => Iterable
...
         currentIterator = plainFun(grab(in)).iterator
-        pushPull()
+        pushPull(pullFirstSubElement = true)
       } catch handleException
 
     override def onUpstreamFinish(): Unit = onFinish()
 
     override def onPull(): Unit =
-      try pushPull()
+      try pushPull(pullFirstSubElement = true)
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2245]: Shouldn't it be `pullFirstSubElement = false` here?

Comment (ygree) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2245]: Good catch! Thanks!

**Review by jrudolph — COMMENTED** [review]
Even if I know what the code should do, I find it hard to compare the code with the expectation of what it should do.

Could the conditions be removed (for a bit of performance hit)?

Then we could always `suspendContext` after an element was grabbed, and always resume/suspend after an element was pushed.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 1 above)
Comment (ygree) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Implemented a test on Telemetry side to catch this issue.

**Comment by ygree** [comment]
> Even if I know what the code should do, I find it hard to compare the code with the expectation of what it should do.
> 
> Could the conditions be removed (for a bit of performance hit)?
> 
> Then we could always `suspendContext` after an element was grabbed, and always resume/suspend after an element was pushed.

It would require an additional suspend / resume pair for each first element. That's why I introduced this condition.

**Review by raboof — APPROVED** [review]
Seems relatively low-risk from the core Akka perspective, but indeed not super easy to follow. If that is well-tested on the Telemetry side then that's perhaps fine?

**Comment by ygree** [comment]
@jrudolph This is an alternative implementation to reduce switch logic: #30290

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -2212,14 +2212,20 @@ private[akka] final class StatefulMapConcat[In, Out](val f: () => In => Iterable
...
     var currentIterator: Iterator[Out] = _
     var plainFun = f()
+    val contextPropagation = ContextPropagation()
 
     def hasNext = if (currentIterator != null) currentIterator.hasNext else false
 
     setHandlers(in, out, this)
 
-    def pushPull(): Unit =
+    def pushPull(pullFirstSubElement: Boolean): Unit =
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2221]: Maybe just rename the flag to `shouldResumeContext`.

Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2221]: Or to keep the original "polarity" of the flag: `usingOriginalContext`.

Then the condition would read `if(!usingOriginalContext) contextPropagation.resumeContext()`

Comment (ygree) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I think `shouldResumeContext` better stands the intention

**Review by jrudolph — APPROVED** [review]
LGTM

**Pull request merged**